Custom error when moving arg outside of its closure

estebank opened the pull request:
When given the following code:

```rust
fn give_any<F: for<'r> FnOnce(&'r ())>(f: F) {
    f(&());
}

fn main() {
    let mut x = None;
    give_any(|y| x = Some(y));
}
```

provide a custom error:

```
error: borrowed data cannot be moved outside of its closure
 --> file.rs:7:27
  |
6 |     let mut x = None;
  |         ----- borrowed data cannot be moved into here...
7 |     give_any(|y| x = Some(y));
  |              ---          ^ cannot be moved outside of its closure
  |              |
  |              ...because it cannot outlive this closure
```

instead of the generic lifetime error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> file.rs:7:27
  |
7 |     give_any(|y| x = Some(y));
  |                           ^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 7:14...
 --> file.rs:7:14
  |
7 |     give_any(|y| x = Some(y));
  |              ^^^^^^^^^^^^^^^
note: ...so that expression is assignable (expected &(), found &())
 --> file.rs:7:27
  |
7 |     give_any(|y| x = Some(y));
  |                           ^
note: but, the lifetime must be valid for the block suffix following statement 0 at 6:5...
 --> file.rs:6:5
  |
6 | /     let mut x = None;
7 | |     give_any(|y| x = Some(y));
8 | | }
  | |_^
note: ...so that variable is valid at time of its declaration
 --> file.rs:6:9
  |
6 |     let mut x = None;
  |         ^^^^^
```

Fix #45983.

estebank committed:
Custom error when moving arg outside of its closure

When given the following code:

```rust
fn give_any<F: for<'r> FnOnce(&'r ())>(f: F) {
    f(&());
}

fn main() {
    let mut x = None;
    give_any(|y| x = Some(y));
}
```

provide a custom error:

```
error: borrowed data cannot be moved outside of its closure
 --> file.rs:7:27
  |
6 |     let mut x = None;
  |         ----- binding declared outside of closure
7 |     give_any(|y| x = Some(y));
  |              ---          ^ cannot be assigned to binding outside of its closure
  |              |
  |              closure you can't escape
```

instead of the generic lifetime error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> file.rs:7:27
  |
7 |     give_any(|y| x = Some(y));
  |                           ^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 7:14...
 --> file.rs:7:14
  |
7 |     give_any(|y| x = Some(y));
  |              ^^^^^^^^^^^^^^^
note: ...so that expression is assignable (expected &(), found &())
 --> file.rs:7:27
  |
7 |     give_any(|y| x = Some(y));
  |                           ^
note: but, the lifetime must be valid for the block suffix following statement 0 at 6:5...
 --> file.rs:6:5
  |
6 | /     let mut x = None;
7 | |     give_any(|y| x = Some(y));
8 | | }
  | |_^
note: ...so that variable is valid at time of its declaration
 --> file.rs:6:9
  |
6 |     let mut x = None;
  |         ^^^^^
```

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

durka commented:
I find the phrasing confusing, especially "closure you can't escape" -- *I* am not in a closure! How about something like this:

```
error: borrowed data cannot be moved outside of its closure
 --> file.rs:7:27
  |
6 |     let mut x = None;
  |         ----- move destination (outside closure)
7 |     give_any(|y| x = Some(y));
  |              ---          ^ borrowed data which cannot escape this closure
  |              |
  |              closure
```

I'm not even sure about the word "escape". Maybe too jargon-ish?

estebank commented:
> _I_ am not in a closure!

After the second glass of wine, I certainly am.

> I'm not even sure about the word "escape". Maybe too jargon-ish?

It probably is, but wanted to push the PR out for exactly this feedback :)

I like your suggested output, but I feel we could still further improve it. I'm just not sure how.

durka commented:
We could do an ellipsis monstrosity:

```
error: borrowed data cannot be moved outside of its closure
 --> file.rs:7:27
  |
6 |     let mut x = None;
  |         ----- borrowed data cannot be moved to here...
7 |     give_any(|y| x = Some(y));
  |              ---          ^ ...from here...
  |              |
  |              ...because it cannot outlive this closure
```

estebank commented:
I didn't want to do that because the closure could be 

```
|y| {
    Some(y)
}
```

durka commented:
Yeah, it would require some kind of special case depending on ordering which isn't great.

estebank commented:
What about

```
error: borrowed data cannot be moved outside of its closure
 --> file.rs:7:27
  |
6 |     let mut x = None;
  |         ----- borrowed data cannot be moved to here...
7 |     give_any(|y| x = Some(y));
  |              ---          ^ cannot be moved outside of its closure
  |              |
  |              ...because it cannot outlive this closure
```

It doesn't read as nicely, but it has the advantage of always making kind of sense, regardless of the code's formatting.

petrochenkov commented:
r? @nikomatsakis

estebank committed:
Reword diagnostic

estebank committed:
Generalize cases where specific move error ocurrs

Trigger new diagnostic in `compile-fail/regions-escape-bound-fn.rs`
test, and not only in `compile-fail/regions-escape-bound-fn-2.rs`.

estebank commented:
There's an incorrectly handled [case](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/issue-7573.rs):

```
% rustc src/test/compile-fail/issue-7573.rs
error: borrowed data cannot be moved outside of its closure
  --> src/test/compile-fail/issue-7573.rs:27:31
   |
27 |     let mut lines_to_use: Vec<&CrateId> = Vec::new();
   |                               ^ cannot be moved outside of its closure
28 |
29 |     let push_id = |installed_id: &CrateId| {
   |         -------   ------------------------ ...because it cannot outlive this closure
   |         |
   |         borrowed data cannot be moved into here...

error: aborting due to previous error
```

It should probably look this way:

```
error: borrowed data cannot be moved outside of its closure
  --> src/test/compile-fail/issue-7573.rs:27:31
   |
27 |     let mut lines_to_use: Vec<&CrateId> = Vec::new();
   |         ---------------- borrowed data cannot be moved into here...
28 |
29 |     let push_id = |installed_id: &CrateId| {
   |                   ------------------------ ...because it cannot outlive this closure
30 |        lines_to_use.push(installed_id);
   |                          ^^^^^^^^^^^^ cannot be moved outside of its closure

error: aborting due to previous error
```

estebank committed:
Handle case of moving into vec with uninferred lifetime

nikomatsakis commented:
@estebank 

> There's an incorrectly handled case:

Hmm, I sort of suspected this would be the case.

nikomatsakis commented:
Oh, I guess you're saying you fixed it. Let me read what's there first. =)

nikomatsakis commented on src/librustc/infer/error_reporting/mod.rs:
```diff
@@ -1067,6 +1067,43 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                        err.span_label(sup_sp, "cannot be moved outside of its closure");
+                        if sup_sp == origin_sp {
+                            err.span_label(*external_span,
+                                           "borrowed data cannot be moved into here...");
```
I still find the word `move` potentially misleading. I would prefer "stored".

estebank replied:
Changed

nikomatsakis commented on src/test/ui/borrowck/issue-45983.stderr:
```diff
@@ -0,0 +1,12 @@
+   |              |
+   |              ...because it cannot outlive this closure
+
+error: aborting due to previous error
```
can you add a `ui` test for the other case you mentioned? The one where you push into a vector that exists outside the closure?

estebank replied:
Moved existing test that shows the behavior to ui: https://github.com/rust-lang/rust/pull/47144/files#diff-e5107c6d2a404ecd34f782e9e9e8b20f

shepmaster commented:
Ping from triage, @estebank! Do you think you'll be able to address some of the feedback here?

estebank committed:
Move diagnostic logic to its own module

- Move specialized borrow checker diagnostic for bindings escaping its
  closure to its own module.
- Move affected tests to `ui`.

estebank commented:
Addressed the review comments.

estebank commented on src/librustc/infer/error_reporting/mod.rs:
```diff
@@ -1067,6 +1067,43 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+
+        // #45983: when trying to assign the contents of an argument to a binding outside of a
+        // closure, provide a specific message pointing this out.
+        if let (&SubregionOrigin::BindingTypeIsNotValidAtDecl(ref external_span),
```
Moved to its own module.

estebank commented on src/test/ui/borrowck/regions-escape-bound-fn.stderr:
```diff
@@ -0,0 +1,12 @@
+   |              |       |    |
+   |              |       |    cannot be stored outside of its closure
+   |              |       cannot infer an appropriate lifetime
+   |              borrowed data cannot outlive this closure
```
@nikomatsakis I'm not entirely happy with the output on this case :-/

This triggers [the case I originally neglected](https://github.com/rust-lang/rust/pull/47144/files#diff-e5107c6d2a404ecd34f782e9e9e8b20f) and was trying to fix...

Should we keep this PR on hold until I have time to look at how to differentiate between these two cases for a more appropriate output? In my mind, this case should completely ignore the lifetime issue for `Some(_)` and point at the destination binding, like in the output of the original case for this PR:

```

error: borrowed data cannot be stored outside of its closure
  --> $DIR/regions-escape-bound-fn.rs:18:27
   |
17 |     let mut x: Option<&isize> = None;
   |            - borrowed data cannot be stored into here...
18 |     with_int(|y| x = Some(y));
   |              ---          ^ cannot be stored outside of its closure
   |              |
   |              ...because it cannot outlive this closure

```

It should be possible to do so as [the current diagnostic](https://play.rust-lang.org/?gist=844fbf39824c9fd8170b6c1eda08121d&version=nightly) does point at `x`'s declaration:

```
error[E0495]: cannot infer an appropriate lifetime for automatic coercion due to conflicting requirements
 --> src/main.rs:8:22
  |
8 |     with_int(|y| x = Some(y));
  |                      ^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the body at 8:14...
 --> src/main.rs:8:14
  |
8 |     with_int(|y| x = Some(y));
  |              ^^^^^^^^^^^^^^^
note: ...so that reference does not outlive borrowed content
 --> src/main.rs:8:27
  |
8 |     with_int(|y| x = Some(y));
  |                           ^
note: but, the lifetime must be valid for the block suffix following statement 0 at 7:5...
 --> src/main.rs:7:5
  |
7 | /     let mut x: Option<&isize> = None;
8 | |     with_int(|y| x = Some(y));
9 | |     //~^ ERROR borrowed data cannot be stored outside of its closure
10| | }
  | |_^
note: ...so that variable is valid at time of its declaration
 --> src/main.rs:7:9
  |
7 |     let mut x: Option<&isize> = None;
  |         ^^^^^
```

Also, should I create a new error code for each of these cases, one shared between all of them or reuse the existing one?

nikomatsakis replied:
I think this output is way better than the original output. The main change I would do is not to print the "cannot infer an appropriate lifetime" label, which seems to add zero value. Saying that the data cannot escape the closure and indicating the closure is pretty awesome.

nikomatsakis commented on src/librustc/infer/error_reporting/nice_region_error/outlives_closure.rs:
```diff
@@ -0,0 +1,94 @@
+                            } else {
+                                err.span_label(*closure_span,
+                                               "borrowed data cannot outlive this closure");
+                                err.span_label(origin_sp,
```
I think we should just drop this label. I don't really think it conveys useful information to the end-user. The `origin_sp` is not *especially* meaningful -- it's basically *a point* where the region is relevant, but not necessarily a particularly good choice of point.

estebank replied:
I feel uneasy about removing the span completely because it's the original error span, and want to point to it in some way. Modified so that it occurs in fewer cases, but will keep it for the #7573 case. I'm planning to revisit that and [the type mismatch cases](https://github.com/estebank/rust/blob/71e0e360111b671bfce65da29b9e06fbcf798872/src/test/ui/closure-expected-type/expect-region-supply-region.stderr) at a later PR.

estebank committed:
Tweak wording and spans of closure lifetime errors

nikomatsakis commented:
Ok, this feels like progress in any case. r=me once travis is happy.

estebank commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 6f9ecaa has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 6f9ecaa7cf0a15db46d13d72932acb2d678c29f1 -->

bors commented:
:hourglass: Testing commit 6f9ecaa7cf0a15db46d13d72932acb2d678c29f1 with merge eabb46a2351fab39802b4ad72bce00c5d83063f5...

bors commented:
:boom: Test timed out

estebank commented:
@bors retry

bors commented:
:hourglass: Testing commit 6f9ecaa7cf0a15db46d13d72932acb2d678c29f1 with merge bc072ed0ca8e2e9f8c79fb04e85b47b5c0e8d6ae...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6100), [status-travis](https://travis-ci.org/rust-lang/rust/builds/331666162?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing bc072ed0ca8e2e9f8c79fb04e85b47b5c0e8d6ae to master...
